libvast/builtins/operators/enumerate.cpp

```diff
@@ -0,0 +1,142 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <vast/arrow_table_slice.hpp>
+#include <vast/concept/parseable/vast/option_set.hpp>
+#include <vast/concept/parseable/vast/pipeline.hpp>
+#include <vast/detail/narrow.hpp>
+#include <vast/error.hpp>
+#include <vast/pipeline.hpp>
+#include <vast/plugin.hpp>
+#include <vast/table_slice.hpp>
+#include <vast/table_slice_builder.hpp>
+
+#include <arrow/type.h>
+
+namespace vast::plugins::enumerate {
+
+namespace {
+
+// TODO: refactor this once we have a modifier that allows for toggling
+// between intra-schema and inter-schema processing.
+class enumerate_operator final : public crtp_operator<enumerate_operator> {
+public:
+  explicit enumerate_operator(std::string field) : field_{std::move(field)} {
+  }
+
+  auto operator()(generator<table_slice> input) const
+    -> generator<table_slice> {
+    // Per-schema state.
+    auto current_type = type{};
+    std::unordered_map<type, uint64_t> offsets;
+    // Create transformation to prepend  column to slice.
+    auto transformations = std::vector<indexed_transformation>{};
+    auto function = [&](struct record_type::field field,
+                        std::shared_ptr<arrow::Array> array) mutable
+      -> std::vector<
+        std::pair<struct record_type::field, std::shared_ptr<arrow::Array>>> {
+      // Create a column with the RIDs.
+      auto n = array->length();
+      auto rid_type = uint64_type{};
+      auto builder
+        = uint64_type::make_arrow_builder(arrow::default_memory_pool());
+      auto reserve_result = builder->Reserve(n);
+      VAST_ASSERT_CHEAP(reserve_result.ok(), reserve_result.ToString().c_str());
+      // Fill the column.
+      auto& offset = offsets[current_type];
+      for (uint64_t i = 0; i < detail::narrow_cast<uint64_t>(n); ++i) {
+        auto append_result
+          = append_builder(rid_type, *builder, view<uint64_t>{offset + i});
+        VAST_ASSERT(append_result.ok(), append_result.ToString().c_str());
+      }
+      offset += n;
+      // Replace first column with a pair of (RID, first).
+      auto rid_array = builder->Finish().ValueOrDie();
+      return {
+        {{field_, rid_type}, rid_array},
+        {field, array},
+      };
+    };
+    transformations.emplace_back(offset{0}, std::move(function));
+    for (auto&& slice : input) {
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      };
+      current_type = slice.schema();
+      co_yield transform_columns(slice, transformations);
```

Comment:
This is UB if the schema already contains a field with the configured name. In that case, you should `ctrl.warn(...)` with a meaningful message and skip the transformation. This message should be emitted at most once per unique schema.